File: raytrace.py
from __future__ import division
from __future__ import print_function

import abc
import argparse
import collections
import functools
import itertools
import math
import multiprocessing
import random
import sys

import numpy
import numpy.linalg

def unit(v):
  return v / numpy.sqrt(numpy.sum(v ** 2))

def dot(a, b):
  return numpy.dot(a,b)

def cross(a, b):
  return numpy.cross(a,b)

class Ray(collections.namedtuple('Ray', 'origin direction')):

  def point_at(self, t):
    return self.origin + t * self.direction


def lerp(a, b, t):
  return (1 - t) * a + t * b

def color(ray, hitable, depth=0):
  record = hitable.hit(ray, 0.001, 100000)
  if record is None:
    t = 0.5 * (unit(ray.direction)[0] + 1.0)
    return lerp(numpy.ones(3), numpy.array([0.5, 0.7, 1.0]), t)
  else:
    if depth > 50:
      return numpy.zeros(3)
    scattered, attenuation = record.material.scatter(ray, record)
    if scattered is None:
      return numpy.zeros(3)
    return attenuation * color(scattered, hitable, depth + 1)


HitRecord = collections.namedtuple('HitRecord', 't pos normal material')

class Hitable(object):
 
  __metaclass__ = abc.ABCMeta

  @abc.abstractmethod
  def hit(self, ray, t_min, t_max):
    "Returns None or a HitRecord if the ray hits between t_min and t_max"


class Sphere(Hitable):

  def __init__(self, center, radius, material):
    self._center = center
    self._radius = radius
    self._material = material
  
  def hit(self, ray, t_min, t_max):
    oc = ray.origin - self._center
    a = dot(ray.direction, ray.direction)
    b = 2 * dot(oc, ray.direction)
    c = dot(oc, oc) - self._radius * self._radius
    roots = numpy.roots([a, b, c])
    real = roots.real[abs(roots.imag)<1e-5]
    for t in numpy.flip(real):
      if t < t_max and t > t_min:
        point = ray.point_at(t)
        normal = (point - self._center) / self._radius
        return HitRecord(t, point, normal, self._material)
    return None

class HitableList(Hitable):

  def __init__(self, objects=()):
    self._objects = list(objects)

  def hit(self, ray, t_min, t_max):
    record = None
    closest = t_max
    for hitable in self._objects:
      this_record = hitable.hit(ray, t_min, closest)
      if this_record is None:
        continue
      closest = this_record.t
      record = this_record
    return record
 
class Camera(object):
  def __init__(self, look_from, look_at, up, vfov, aspect, aperture, focus_dist):
    theta = vfov * math.pi / 180.
    half_height = math.tan(theta / 2)
    half_width = aspect * half_height
    w = unit(look_from - look_at)
    u = unit(cross(up, w))
    v = cross(w, u)
    self._origin = look_from
    self._radius = aperture / 2
    self._lower_left = look_from - half_width * focus_dist * u - half_height * focus_dist * v - focus_dist * w
    self._horizontal = 2 * half_width * focus_dist * u
    self._vertical = 2 * half_height * focus_dist * v
    self._w = w
    self._u = u
    self._v = v

  def get_ray(self, u, v):
    rd = self._radius * random_in_unit_disk()
    offset  = self._u * rd[0] + self._v * rd[1]
    return Ray(self._origin + offset, self._lower_left + u * self._horizontal + v * self._vertical - self._origin - offset)


def random_in_unit_disk():
  while True:
    p = 2 * numpy.array([random.random(), random.random(), 0]) - numpy.array([1, 1, 0])
    if dot(p, p) < 1.0:
      return p

def random_in_unit_sphere():
  while True:
    p = 2 * numpy.random.random(3) - numpy.ones(3)
    if numpy.sum(p ** 2) < 1:
      return p


class Material(object):

  __metaclass__ = abc.ABCMeta

  @abc.abstractmethod
  def scatter(self, ray, hit_record):
    "Return the attenuation and scattered ray, or None if it is absorbed"


class Lambertian(Material):

  def __init__(self, albedo):
    self._albedo = albedo

  def scatter(self, ray, hit_record):
    target = hit_record.pos + hit_record.normal + random_in_unit_sphere()
    scattered = Ray(hit_record.pos, target - hit_record.pos)
    return scattered, self._albedo

def reflect(vector, normal):
  return vector - 2 * dot(vector, normal) * normal


class Metal(Material):
  def __init__(self, albedo, fuzz):
    self._albedo = albedo
    self._fuzz = fuzz

  def scatter(self, ray, hit_record):
    reflected = reflect(unit(ray.direction), hit_record.normal)
    scattered = Ray(hit_record.pos, reflected + self._fuzz * random_in_unit_sphere())
    if dot(scattered.direction, hit_record.normal) > 0:
      return scattered, self._albedo
    return None, None


def refract(vector, normal, ni_over_nt):
  uv = unit(vector)
  dt = dot(uv, normal)
  discriminant = 1.0 - ni_over_nt * ni_over_nt * (1 - dt * dt)
  if discriminant > 0:
    return ni_over_nt * (uv - normal * dt) - normal * math.sqrt(discriminant)
  else:
    return None

def schlick(cosine, ref_idx):
  r0 = (1 - ref_idx) / (1 + ref_idx)
  r0 *= r0
  return r0 + (1-r0) * ((1 - cosine) ** 5)

class Dialectric(Material):
  def __init__(self, refraction):
    self._refraction = refraction

  def scatter(self, ray, hit_record):
    if dot(ray.direction, hit_record.normal) > 0:
      outward_normal = -hit_record.normal
      ni_over_nt = self._refraction
      cosine = self._refraction * dot(ray.direction, hit_record.normal) / length(ray.direction)
    else:
      outward_normal = hit_record.normal
      ni_over_nt = 1 / self._refraction
      cosine = -dot(ray.direction, hit_record.normal) / length(ray.direction)
    refracted = refract(ray.direction, outward_normal, ni_over_nt)
    if refracted is None:
      reflect_prob = 1.0
    else:
      reflect_prob = schlick(cosine, self._refraction)
    if random.random() < reflect_prob:
      reflected = reflect(ray.direction, hit_record.normal)
      return Ray(hit_record.pos, reflected), numpy.ones(3)
    else:
      return Ray(hit_record.pos, refracted), numpy.ones(3)
    

def render((y, x), width, height, camera, objects, samples):
  c = numpy.array([0., 0., 0.])
  for _ in range(samples):
    u = (x + random.random()) / width
    v = (y + random.random()) / height
    r = camera.get_ray(u, v)
    c += color(r, objects)
  c /= samples
  c = numpy.sqrt(c)
  return 255.9 * c


def length(v):
  return numpy.sqrt(numpy.sum(v**2))


def parse_args():
  parser = argparse.ArgumentParser()
  parser.add_argument('--width', type=int, default=640)
  parser.add_argument('--height', type=int, default=480)
  parser.add_argument('--samples', type=int, default=10)
  parser.add_argument('--workers', type=int, default=8)
  return parser.parse_args()


def main():
  args = parse_args()
  width, height = args.width, args.height
  print("P3")
  print(width, height)
  print(255)

  look_from = numpy.array([-3., 3., 2.])
  look_at = numpy.array([0., 0., -1.])
  camera = Camera(look_from, look_at, numpy.array([0., 1.0, 0.]), 30, width / height, 2.0, length(look_from - look_at))

  objects = HitableList([
     Sphere(numpy.array([0.0, 0.0, -1.0]), 0.5, Lambertian(numpy.array([0.3, 0.6, 0.3]))),
     Sphere(numpy.array([0.0, -100.5, -1.0]), 100.0, Lambertian(numpy.array([0.8, 0.8, 0.0]))),
     Sphere(numpy.array([1.0, 0.0, -1.0]), 0.5, Metal(numpy.array([0.8, 0.6, 0.2]), 0.3)),
     Sphere(numpy.array([-1.0, 0.0, -1.0]), 0.5, Dialectric(1.5)),
     Sphere(numpy.array([-1.0, 0.0, -1.0]), -0.45, Dialectric(1.5)),
  ])

  pool = multiprocessing.Pool(args.workers)
  pixels = itertools.product(range(height-1, -1, -1), range(width))

  image = pool.map(
      functools.partial(render, width=width, height=height,
                                camera=camera, objects=objects,
                                samples=args.samples),
      pixels)
  for pixel in image:
    print(int(pixel[0]), int(pixel[1]), int(pixel[2]))


if __name__ == '__main__':
  main()

File: raytrace.hs
import Data.Maybe
import System.Random

-- Vec3 operations

data Vec3 = Vec3 { x :: Float, y :: Float, z :: Float } deriving (Show)
type Image = [Vec3]

add :: Vec3 -> Vec3 -> Vec3
add v1 v2 = Vec3 (x v1 + x v2) (y v1 + y v2) (z v1 + z v2)

negate3 :: Vec3 -> Vec3
negate3 v = Vec3 (-x v) (-y v) (-z v)

multiply :: Vec3 -> Vec3 -> Vec3
multiply v1 v2 = Vec3 (x v1 * x v2) (y v1 * y v2) (z v1 * z v2)

divide :: Vec3 -> Vec3 -> Vec3
divide v1 v2 = Vec3 (x v1 / x v2) (y v1 / y v2) (z v1 / z v2)

scale :: Vec3 -> Float -> Vec3
scale v t = Vec3 (t * x v) (t * y v) (t * z v)

dot :: Vec3 -> Vec3 -> Float
dot v1 v2 = (x v1) * (x v2) + (y v1) * (y v2) + (z v1) * (z v2)

cross :: Vec3 -> Vec3 -> Vec3
cross v1 v2 = Vec3 ((y v1) * (z v2) - (z v1) * (y v2))
                   ((z v1) * (x v2) - (x v1) * (z v2))
                   ((x v1) * (y v2) - (y v1) * (x v2))

squared_length :: Vec3 -> Float
squared_length v = (x v) * (x v) + (y v) * (y v) + (z v) * (z v)

len = sqrt . squared_length

normalize :: Vec3 -> Vec3
normalize v = scale v (1 / len v)

lerp :: Vec3 -> Vec3 -> Float -> Vec3
lerp x y t = add (scale x (1-t)) (scale y t)

average3_ :: [Vec3] -> Vec3 -> Int -> Vec3
average3_ [] a n = scale a (1.0 / fromIntegral n)
average3_ (x:xs) a n = average3_ xs (add a x) (n+1)

average3 :: [Vec3] -> Vec3
average3 [] = Vec3 0.0 0.0 0.0
average3 xs = average3_ xs (Vec3 0.0 0.0 0.0) 0

-- Ray operations

data Ray = Ray { origin :: Vec3
               , direction :: Vec3 } deriving (Show)

pointAt :: Ray -> Float -> Vec3
pointAt r t = add (origin r) (scale (direction r) t)

-- Collision Physics

data Hitable = Sphere { center :: Vec3 , radius :: Float , material :: Material}

data HitRecord = HitRecord { time :: Float, position :: Vec3, normal :: Vec3 , material_ :: Material }

hit :: Hitable -> Ray -> Float -> Float -> Maybe HitRecord
hit s r tMin tMax = let oc = add (origin r) (negate3 (center s))
                        a = dot (direction r) (direction r) 
                        b = dot oc (direction r)
                        c = (dot oc oc) - (radius s) * (radius s)
                        discriminant = b * b - a * c 
                    in if discriminant > 0
                       then let tmp = sqrt(b * b - a * c)
                                sol1 = (-b - tmp) / a
                                sol2 = (-b + tmp) / a
                            in if sol1 < tMax && sol1 > tMin
                               then let p = pointAt r sol1
                                        normal = scale (add p (negate3 (center s))) (1.0 / (radius s))
                                    in Just (HitRecord sol1 p normal (material s))
                               else if sol2 < tMax && sol2 > tMin
                                    then let p = pointAt r sol2
                                             normal = scale (add p (negate3 (center s))) (1.0 / (radius s))
                                         in Just (HitRecord sol2 p normal (material s))
                                    else Nothing
                       else Nothing

closestTime :: Float -> Maybe HitRecord -> Float
closestTime t Nothing = t
closestTime t (Just x) = min t (time x)

closestHit :: Maybe HitRecord -> Maybe HitRecord -> Maybe HitRecord
closestHit Nothing Nothing = Nothing
closestHit (Just x) Nothing = (Just x)
closestHit Nothing (Just x) = (Just x)
closestHit (Just x) (Just y) = if (time x) < (time y) then (Just x) else (Just y)

getClosestHit :: [Hitable] -> Ray -> Float -> Float -> Maybe HitRecord
getClosestHit [] _ _ _ = Nothing
getClosestHit (x:xs) r tMin tMax = let this = (hit x r tMin tMax)
                                   in if isNothing this
                                      then getClosestHit xs r tMin tMax
                                   else let other = getClosestHit xs r tMin (closestTime tMax this)
                                        in closestHit this other

-- Image Generation

coords :: StdGen -> Integer -> Integer -> Integer -> [(StdGen, Integer, Integer)]
coords g x nx ny = if x == nx
                   then if ny == 0
                        then []
                        else coords (fst $ split g) 0 nx (ny-1)
                   else ((g, x, ny):coords (fst $ split g) (x+1) nx ny)

toCoord :: Integer -> Integer -> Float -> Float
toCoord x nx u = ((fromIntegral x) + u) / (fromIntegral nx)

genImage :: StdGen -> Integer -> Integer -> [Hitable] -> Image
genImage g nx ny world = [ let us = randoms (snd $ split g') :: [Float]
                               vs = randoms (snd $ split (snd $ split g')) :: [Float]
                     in average3 [ color (getRay camera (toCoord x nx u) (toCoord y ny v)) g' world | (u, v) <- take 10 (zip us vs) ]
                   | (g', x, y) <- coords g 0 nx ny ];

strVec3 :: Vec3 -> String
strVec3 v = show (floor (x v)) ++ " " ++ show (floor (y v)) ++ " " ++ show (floor (z v))

strImage :: Image -> String
strImage [] = ""
strImage (x:xs) = strVec3 x ++ "\n" ++ strImage xs

-- Materials

data Material = Lambertian { albedo :: Vec3 }
              | Metal { albedo :: Vec3 , fuzz :: Float }
              | Dialectric { ri :: Float }

randomInUnitSphere :: StdGen -> (Vec3, StdGen)
randomInUnitSphere g = let range = (-1.0, 1.0)
                           (x, g') = randomR range g
                           (y, g'') = randomR range g'
                           (z, g''') = randomR range g''
                           v = Vec3 x y z
                       in if squared_length v < 1.0 then (v, g''') else randomInUnitSphere g'''

reflect :: Vec3 -> Vec3 -> Vec3
reflect v n = add v (scale n (-2 * dot v n))

refract :: Vec3 -> Vec3 -> Float -> Maybe Vec3
refract v n ni_over_nt = let uv = normalize v
                             dt = dot uv n
                             discriminant = 1.0 - ni_over_nt * ni_over_nt * (1 - dt * dt)
                         in if discriminant > 0
                            then let a = scale (add uv (scale n (-1.0 * dt))) ni_over_nt
                                     b = scale n (-1.0 * sqrt discriminant)
                                 in Just(add a b)
                            else Nothing

schlick :: Float -> Float -> Float
schlick c i = let r0 = (1.0 - i) / (1.0 + i)
                  r1 = r0 * r0
              in r1 + (1.0 - r1) * ((1.0 - c) ** 5.0)

glass :: StdGen -> HitRecord -> (Maybe Vec3) -> Vec3 -> Float -> Float -> (Maybe (Vec3, Ray), StdGen)
glass g hr Nothing reflected _ _ = (Just (Vec3 1.0 1.0 1.0, Ray (position hr) reflected), g)
glass g hr (Just refracted) reflected cosine i = let (ref, g') = random g
                                                     dir = if ref < schlick cosine i then reflected else refracted
                                                 in (Just (Vec3 1.0 1.0 1.0, Ray (position hr) dir), g')

scatter :: StdGen -> Material -> Ray -> HitRecord -> (Maybe (Vec3, Ray), StdGen)
scatter g (Lambertian a) r hr = let (u, g') = randomInUnitSphere g
                                    target = add (add (position hr) (normal hr)) u
                                    scattered = Ray (position hr) (add target (negate3 (position hr)))
                                in (Just (a, scattered), g')
scatter g (Metal a f) r hr = let reflected = reflect (normalize $ direction r) (normal hr)
                                 (u, g') = randomInUnitSphere g
                                 scattered = Ray (position hr) (add reflected (scale u f))
                                 d = dot reflected (normal hr)
                             in (if d > 0 then Just (a, scattered) else Nothing, g')
scatter g (Dialectric i) r hr = let reflected = reflect (direction r) (normal hr)
                                    discriminant = dot (direction r) (normal hr)
                                    outward_normal = if discriminant > 0
                                                     then scale (normal hr) (-1.0)
                                                     else normal hr
                                    ni_over_nt = if discriminant > 0 then i else 1.0 / i
                                    cosine = if discriminant > 0
                                             then i * (dot (direction r) (normal hr)) / len (direction r)
                                             else (-1.0) * (dot (direction r) (normal hr)) / len (direction r)
                                    refracted = refract (direction r) outward_normal ni_over_nt
                                in glass g hr refracted reflected cosine i

-- Rendering Colors

render :: StdGen -> Maybe (Vec3, Ray) -> [Hitable] -> Integer -> (Vec3, StdGen)
render g Nothing _ _ = (Vec3 0.0 0.0 0.0, g)
render g (Just (v, r)) world d = let (c, g') = color_ r g world (d-1)
                                 in (multiply v c, g')

hitColor :: StdGen -> (Maybe HitRecord) -> Ray -> [Hitable] -> Integer -> (Vec3, StdGen)
hitColor g _ _ _ 0 = (Vec3 0.0 0.0 0.0, g)
hitColor g Nothing r _ _ = (lerp (Vec3 1.0 1.0 1.0) (Vec3 0.5 0.7 1.0) (0.5 * ((y (normalize (direction r))) + 1)), g)
hitColor g (Just rec) r w d = let (scattered, g') = scatter g (material_ rec) r rec
                              in render g' scattered w d

color_ :: Ray -> StdGen -> [Hitable] -> Integer -> (Vec3, StdGen)
color_ r g world d = hitColor g (getClosestHit world r 0.00001 1000000) r world d

color :: Ray -> StdGen -> [Hitable] -> Vec3
color r g world = let (v, _) = color_ r g world 50
                    in scale (Vec3 (sqrt $ x v) (sqrt $ y v) (sqrt $ z v)) 255.99

data Camera = Camera { lower_left :: Vec3, horizontal :: Vec3, vertical :: Vec3, imgOrigin :: Vec3 }
camera = Camera (Vec3 (-2.0) (-1.0) (-1.0)) (Vec3 4.0 0.0 0.0) (Vec3 0.0 2.0 0.0) (Vec3 0.0 0.0 0.0)

getRay :: Camera -> Float -> Float -> Ray
getRay c u v = Ray (imgOrigin c) (add (lower_left c) (add (scale (horizontal c) u) (add (scale (vertical c) v) (negate3 (imgOrigin c)))))

rands :: StdGen -> [StdGen]
rands g = let (g', g'') = split g in (g':rands g'')

main = do {
   putStrLn "P3";
   putStrLn "400 200";
   putStrLn "255";
   stdGen <- getStdGen;
   world <- return [(Sphere (Vec3 0.0 0.0 (-1.0)) 0.5 (Lambertian $ Vec3 0.8 0.3 0.3)),
                    (Sphere (Vec3 0.0 (-100.5) (-1.0)) 100.0 (Lambertian $ Vec3 0.8 0.8 0.0)),
                    (Sphere (Vec3 1.0 0.0 (-1.0)) 0.5 (Metal (Vec3 0.8 0.6 0.2) 0.3)),
                    (Sphere (Vec3 (-1.0) 0.0 (-1.0)) 0.5 (Dialectric 1.5)),
                    (Sphere (Vec3 (-1.0) 0.0 (-1.0)) (-0.45) (Dialectric 1.5))];
   putStrLn $ strImage $ genImage stdGen 400 200 world ;
}
